lib/compress/zstd_compress.c

```diff
@@ -2853,14 +2853,23 @@ static size_t ZSTD_checkDictNCount(short* normalizedCounter, unsigned dictMaxSym
 
 size_t ZSTD_loadCEntropy(ZSTD_compressedBlockState_t* bs, void* workspace,
                          short* offcodeNCount, unsigned* offcodeMaxValue,
-                         const void* const dict, size_t dictSize) 
+                         const void* const dict, size_t dictSize)
 {
     const BYTE* dictPtr = (const BYTE*)dict;    /* skip magic num and dict ID */
     const BYTE* const dictEnd = dictPtr + dictSize;
     dictPtr += 8;
 
     {   unsigned maxSymbolValue = 255;
-        size_t const hufHeaderSize = HUF_readCTable((HUF_CElt*)bs->entropy.huf.CTable, &maxSymbolValue, dictPtr, dictEnd-dictPtr);
+        size_t hasNoZeroWeights;
+        size_t const hufHeaderSize = HUF_readCTable((HUF_CElt*)bs->entropy.huf.CTable, &maxSymbolValue, dictPtr,
+            dictEnd-dictPtr, &hasNoZeroWeights);
+
+        /* We only set the loaded table as valid if it contains all non-zero
+         * weights. Otherwise, we set it to check */
+        if (hasNoZeroWeights)
+            bs->entropy.huf.repeatMode = HUF_repeat_valid;
+        else bs->entropy.huf.repeatMode = HUF_repeat_check;
```

Comment:
Maybe this should be HUF_repeat_none instead of HUF_repeat_check? Since the dictionary is 'corrupt' at this point right?

Reply:
It depends on if we want to forbid such cases (zero-weighted symbols in the dictionary header) or we want to authorize them.

At this point, the specification doesn't forbid it, and I believe it's correct.

We can imagine a use case where a dictionary is employed to compress ASCII files, in which case many byte values are simply impossible. It then becomes advantageous to zero-weight the impossible byte values, for better compression.

Of course, if the file to compress contains a few non-ASCII characters, the assumption doesn't work anymore, and the compressed payload is penalized by the requirement to define its own table. But since it's _not_ supposed to happen, it's okay to shift this cost onto bogus exceptions.

Reply:
I agree we should allow it.